### PR TITLE
8265604: Support unlinked classes in dynamic CDS archive

### PR DESCRIPTION
**Pull request by calvinccheung**
This change adds the capability of storing unlinked class in dynamic CDS archive.
Before this change, unlinked classes loaded by custom class loaders won't be store in dynamic CDS archive.
Unlinked old classes with version < JAVA_6_VERSION will also be supported by this change. Similar to the support for old classes in static CDS archive in [JDK-8261090](https://bugs.openjdk.java.net/browse/JDK-8261090).

Testing:

- [x] tiers1,2,3,4

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265604](https://bugs.openjdk.java.net/browse/JDK-8265604): Support unlinked classes in dynamic CDS archive

### Reviewers
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4803/head:pull/4803` \
`$ git checkout pull/4803`

Update a local copy of the PR: \
`$ git checkout pull/4803` \
`$ git pull https://git.openjdk.java.net/jdk pull/4803/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4803`

View PR using the GUI difftool: \
`$ git pr show -t 4803`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4803.diff">https://git.openjdk.java.net/jdk/pull/4803.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 1]
```diff
@@ -571,10 +571,26 @@ class CollectCLDClosure : public CLDClosure {
...
+    return false;
+  }
+  if (DynamicDumpSharedSpaces && ik->is_shared_unregistered_class()) {
+    // Linking of unregistered classes at this stage may cause more
+    // classes to be resolved, resulting in calls to ClassLoader.loadClass()
+    // that may not be expected by custom class loaders.
+    //
+    // It's OK to do this for the built-in loaders as we know they can
+    // tolerate this. (Note that unregistered classes are loaded by the NULL
+    // loader during DumpSharedSpaces).
```
Comment (yminqi) [inline src/hotspot/share/cds/metaspaceShared.cpp:590]: DumpSharedSpace -> dump time. (All classes are loaded by boot class loader at dump time)

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 2]
```diff
@@ -934,7 +937,8 @@ bool InstanceKlass::link_class_impl(TRAPS) {
       // 1) the class is loaded by custom class loader or
       // 2) the class is loaded by built-in class loader but failed to add archived loader constraints
       bool need_init_table = true;
-      if (is_shared() && SystemDictionaryShared::check_linking_constraints(THREAD, this)) {
+      if (is_shared() && verified_at_dump_time() &&
```
Comment (yminqi) [inline src/hotspot/share/oops/instanceKlass.cpp:940]: Should comment be updated for the change?

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/CustomLoaderApp.java [hunk 3]
```diff
@@ -36,20 +36,28 @@ public static void main(String args[]) throws Exception {
...
                                        ", expected == " + urlClassLoader);
         }
 
-        Method method = c.getMethod("main", String[].class);
-        String[] params = null;
-        method.invoke(null, (Object)params);
+        if (init) {
+            Object o = c.newInstance();
+
+
```
Comment (yminqi) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/CustomLoaderApp.java:57]: extra empty line.

**Review by yminqi — COMMENTED** [review]
Looks good. Small nits.

File: src/hotspot/share/cds/metaspaceShared.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/cds/metaspaceShared.cpp:590]: During dynamic dump, an unregistered class could be loaded by a custom loader.

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 2 above)
Comment (calvinccheung) [reply src/hotspot/share/oops/instanceKlass.cpp:940]: I will update the comment to the following:
```
      // 1) the class is loaded by custom class loader or
      // 2) the class is loaded by built-in class loader but failed to add archived loader constraints or
      // 3) the class was not verified during dump time
```

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/CustomLoaderApp.java (hunk 3 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/CustomLoaderApp.java:57]: I'll delete the empty lines.
Thanks for your review.

**Review by iklam — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/cds/metaspaceShared.cpp (hunk 1 above)
Comment (yminqi) [reply src/hotspot/share/cds/metaspaceShared.cpp:590]: Just recommend changing DumpSharedSpace -> dump time.

**Review by yminqi — APPROVED** [review]
LGTM. Only a recommendation of  "during DumpSharedSpaces -> during dump time". It is up to you to change it.

**Comment by calvinccheung** [comment]
> LGTM. Only a recommendation of "during DumpSharedSpaces -> during dump time". It is up to you to change it.

Leaving the comment as is is clearer because the comment applies to static dump only and DumpSharedSpaces implies static dump.

**Comment by calvinccheung** [comment]
@iklam, @yminqi Thanks for the review.
/integrate

**Pull request closed without merging**